gufe/protocols/protocoldag.py

```diff
@@ -0,0 +1,64 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/gufe
+
+import abc
+from typing import Iterable, List, Dict, Set, Optional
+
+import networkx as nx
+
+from .protocolunit import ProtocolUnit
+from .results import ProtocolDAGResult
+
+
+class ProtocolDAG:
+    """An executable directed, acyclic graph (DAG) composed of `ProtocolUnit`s
+    with dependencies specified.
+
+    This is the unit of execution passed to an alchemical `Scheduler`.
+    A `ProtocolDAG` yields a `ProtocolResult`, which can be placed in a `ResultStore`.
+
+    """
+
+    def __init__(
+        self,
+        graph: nx.DiGraph,
+        name: Optional[str] = None,
+    ):
+        """Create a new `ProtocolDAG`.
+
+        Parameters
+        ----------
+        graph : nx.DiGraph
+            The `ProtocolUnit`s, with dependencies set, as a networkx `DiGraph`.
+        name : str
+            Unique identifier for this `ProtocolDAG`.
+
+        """
+        self._graph = graph
+        self._name = name
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def graph(self):
+        return self._graph
+
+    def execute(self) -> ProtocolDAGResult:
+        """Execute the full DAG in-serial, in process."""
+        # operate on a copy, since we'll add ProtocolUnitResults as node attributes
+        graph = self._graph.copy(as_view=False)
+
+        # iterate in DAG order
+        for unit in reversed(list(nx.topological_sort(self._graph))):
```

Comment:
So we're having to reverse the provided iterator for DAGs from Networkx.  Not a drama, but it might be worth revisiting if we've got our DAG definition the right way round.

Reply:
Apologies @richardjgowers, from our discussion Friday we had decided to encode DAGs as `ProtocolUnit`s pointing to their dependents, but traversal appeared easier to go the other direction (`ProtocolUnit`s point to their dependencies). 

I should have mentioned this in a comment here that I reversed it in the tests, but I think you noticed this.

Reply:
Either is fine, you're either saying "A leads to B" or "B requires A", I just don't want to be going against the grain with how everyone else seems to phrase it.  Can leave this as-is, but if we keep finding counter examples we might have to switch polarity.